src/deploy/functions/iambinder.ts

```diff
@@ -0,0 +1,298 @@
+    const diff = this.clone();
+    for (const { role, members, condition } of base.allBindings()) {
+      const conditionKey = condition ? IamBindings.getConditionKey(condition) : "";
+      if (diff.additions[role]?.[conditionKey]) {
+        for (const member of members) {
+          diff.additions[role][conditionKey].members.delete(member);
+        }
+        if (diff.additions[role][conditionKey].members.size === 0) {
+          delete diff.additions[role][conditionKey];
+        }
+        if (Object.keys(diff.additions[role]).length === 0) {
+          delete diff.additions[role];
+        }
+      }
+    }
+    return diff;
+  }
+
+  merge(other: IamBindings): IamBindings {
+    const merged = new IamBindings();
+    for (const bindings of [this, other]) {
+      for (const binding of bindings.allBindings()) {
+        merged.add(binding.role, Array.from(binding.members), binding.condition);
+      }
+    }
+    return merged;
+  }
+
+  clone(): IamBindings {
+    const copied = new IamBindings();
+    for (const binding of this.allBindings()) {
+      copied.add(binding.role, Array.from(binding.members), binding.condition);
+    }
+    return copied;
+  }
+
+  asIamBindings(): iam.Binding[] {
+    const iamBindings: iam.Binding[] = [];
+    for (const binding of this.allBindings()) {
+      const copy: iam.Binding = { role: binding.role, members: Array.from(binding.members) };
+      if (binding.condition) {
+        copy.condition = { ...binding.condition };
+      }
+      iamBindings.push(copy);
+    }
+    return iamBindings;
+  }
+
+  private *allBindings(): Generator<IamBinding> {
+    for (const role of Object.keys(this.additions) as Role[]) {
+      for (const conditionKey of Object.keys(this.additions[role])) {
+        yield this.additions[role][conditionKey];
+      }
+    }
+  }
+
+  private static getConditionKey(condition: iam.Binding["condition"]): string {
+    return createHash("md5").update(JSON.stringify(condition)).digest("base64");
+  }
+
+  static fromIamBindings(bindings: iam.Binding[]): IamBindings {
+    const bindingInternal = new IamBindings();
+    for (const { role, members, condition } of bindings) {
+      if (!isRole(role)) {
+        throw new Error(`Invalid role. Role ${role} must be prefixed with "roles/"`);
+      }
+      bindingInternal.add(role, members, condition);
+    }
+    return bindingInternal;
+  }
+}
+
+/**
+ * IamBinder generates necessary IAM binding given different types of function endpoints.
+ */
+export class IamBinder {
+  additions: Record<Resource, IamBindings>; // resource -> bindings
+  storageServiceAccount: string | undefined;
+
+  constructor(private readonly projectId: string, private readonly projectNumber: string) {
+    this.additions = {};
+  }
+
+  async apply(): Promise<iam.Policy[]> {
+    const updates = Object.entries(this.additions).map(async ([resource, bindings]) => {
+      return await IamBinder.updatePolicy(resource as Resource, bindings);
```

Comment:
Are we now calling `getIamPolicy` & `setIamPolicy` for each binding addition? Or did I misunderstand this?

Reply:
This is meant to iterate over all "resources" - e.g. secret manager secret. Adding roles to service account is bit of an outlier in that the "resource" is the GCP project itself.

Reply:
oh resource string _is_ different per iteration. e.g.

```js
{
  '//cloudresourcemanager.googleapis.com/projects/12345': [
    { role: "roles/someRole", members: ["a", "b"] },
    { role: "roles/someRole", members: ["a", "b"] },
  ],

  '//secretmanager.googleapis.com/projects/12345/secrets/MY_SECRET': [
    { role: "roles/secretAccessor", members: ["c"] },
  ],

  '//secretmanager.googleapis.com/projects/12345/secrets/ANOTHER_SECRET': [
    { role: "roles/secretAccessor", members: ["d"] },
  ],

  // ... in the future we may add support for other resources e.g.
 '//cloudfunctions.googleapis.com/projects/12345/locations/us-central1/functions/myFunc': [
    { role: "roles/invoker", members: ["e", "f"] },
  ]
}
```

I'm playing on the idea that there is distinct IAM policy per GCP resource (referred by its [full resource name[(https://cloud.google.com/iam/docs/full-resource-names)).

I was having fun w/ types to help identify different types of resource we support (see `ResourceManagerResource`, `SecretManagerResource`, and `Resource` type definitions) but I'm getting the feeling that this is maybe too "clever"?

Reply:
I think that it might be too clever. Maybe we can just take an API and a resource name.